### PR TITLE
win_dsc - Add argument validation and other fixes

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
Refactor of the win_dsc module that adds the following;

* Parameter validation
* Parameter invocation return
* Return Verbose logs for both the Test and Set methods
* Support for Hashtables/MSFT_KeyValuePair values
* Validation on sub dicts (CimInstances)
* Revamp of the tests to be simpler and touch less system resources

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_dsc

### CODE REVIEW

File: docs/docsite/rst/user_guide/windows_dsc.rst [hunk 1]
```diff
@@ -97,7 +97,68 @@ This is what the Ansible task version of the above DSC Registry resource would l
         Ensure: Present
         Key: HKEY_LOCAL_MACHINE\SOFTWARE\ExampleKey
         ValueName: TestValue
-        ValueData: TestData
+        ValueData: TestDatai
```
Comment (acozine) [inline docs/docsite/rst/user_guide/windows_dsc.rst]:
```suggestion
        ValueData: TestData
```

File: docs/docsite/rst/user_guide/windows_dsc.rst [hunk 2]
```diff
@@ -97,7 +97,68 @@ This is what the Ansible task version of the above DSC Registry resource would l
...
+
+In Ansible 2.8, the ``win_dsc`` module will now automatically validate the
```
Comment (acozine) [inline docs/docsite/rst/user_guide/windows_dsc.rst]:
```suggestion
Starting in Ansible 2.8, the ``win_dsc`` module automatically validates the
```

File: docs/docsite/rst/user_guide/windows_dsc.rst [hunk 3]
```diff
@@ -97,7 +97,68 @@ This is what the Ansible task version of the above DSC Registry resource would l
...
+input options from Ansible with the DSC definition. This means Ansible will
+fail if the option name is incorrect, a mandatory option is not set, or the
+value is not a valid choice. When running Ansible with a verbosity level or 3
```
Comment (acozine) [inline docs/docsite/rst/user_guide/windows_dsc.rst]:
```suggestion
value is not a valid choice. When running Ansible with a verbosity level of 3
```

File: docs/docsite/rst/user_guide/windows_dsc.rst [hunk 4]
```diff
@@ -97,7 +97,68 @@ This is what the Ansible task version of the above DSC Registry resource would l
...
+            "[SERVER2016]: LCM:  [ End    Test     ]  [[Registry]DirectResourceAccess] False in 0.2510 seconds.",
+            "[SERVER2016]: LCM:  [ End    Set      ]    in  0.3310 seconds.",
+            "Operation 'Invoke CimMethod' complete.",
+            "Time taken for configuration job to complete is 0.475 seconds"
+        ]
+    }
+
+The ``invocation.module_args`` key shows the actual values that were set as
+well as other possible values that were not set. Unfortunately this will not
+show the default value for a DSC property just what was set from the Ansible
```
Comment (acozine) [inline docs/docsite/rst/user_guide/windows_dsc.rst]:
```suggestion
show the default value for a DSC property, only what was set from the Ansible
```

File: docs/docsite/rst/user_guide/windows_dsc.rst [hunk 5]
```diff
@@ -97,7 +97,68 @@ This is what the Ansible task version of the above DSC Registry resource would l
...
+task. Any ``*_password`` option will be masked in the output for security
+reasons, if there are any other sensitive module options set ``no_log: True``
```
Comment (acozine) [inline docs/docsite/rst/user_guide/windows_dsc.rst]:
```suggestion
reasons, if there are any other sensitive module options, set ``no_log: True``
```

File: docs/docsite/rst/user_guide/windows_dsc.rst [hunk 6]
```diff
@@ -97,7 +97,68 @@ This is what the Ansible task version of the above DSC Registry resource would l
...
+on the task to stop it from being logged.
```
Comment (acozine) [inline docs/docsite/rst/user_guide/windows_dsc.rst]:
```suggestion
on the task to stop all task output from being logged.
```

File: lib/ansible/modules/windows/win_dsc.py [hunk 7]
```diff
@@ -146,9 +159,24 @@
...
     sample: true
-message:
-    description: Any error message from invoking the DSC resource.
-    returned: error
-    type: str
-    sample: Multiple DSC modules found with resource name xyz
+verbose_test:
+    description: The verbose output as a list from executing the DSC test
+      method.
+    returned: Ansible verbosity if -vvv or greater
```
Comment (acozine) [inline lib/ansible/modules/windows/win_dsc.py]:
```suggestion
    returned: Ansible verbosity is -vvv or greater
```

**Review by acozine — COMMENTED** [review]
Docs look good - I made a couple of small suggestions.

**Comment by jborean93** [comment]
Thanks @acozine for the review, should be fixed.

**Comment by jborean93** [comment]
Merging this in so we get something running in devel.

**Pull request merged**